infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -364,15 +364,58 @@ protected String getNamespaceName(Workspace workspace) throws InfrastructureExce
               EnvironmentContext.getCurrent().getSubject().getUserId(),
               EnvironmentContext.getCurrent().getSubject().getUserName());
       namespace = evaluateLegacyNamespaceName(resolutionCtx);
+
+      LOG.warn(
+          "Workspace '{}' doesn't have an explicit namespace assigned."
+              + " The legacy namespace resolution resolved it to '{}'.",
+          workspace.getId(),
+          namespace);
     }
+
+    if (!NamespaceNameValidator.isValid(namespace)) {
+      // At a certain unfortunate past version of Che, we stored invalid namespace names.
+      // At this point in time, we're trying to work with an existing workspace that never could
+      // started OR has been running since before that unfortunate version. In both cases, going
+      // back to the default namespace name is the most safe bet we can make.
+
+      // but of course, our attempt will be futile if we're running in a context that doesn't know
+      // the current user.
+      Subject subj = EnvironmentContext.getCurrent().getSubject();
+      if (!subj.isAnonymous()) {
+        NamespaceResolutionContext resolutionCtx =
+            new NamespaceResolutionContext(workspace.getId(), subj.getUserId(), subj.getUserName());
+
+        String defaultNamespace = evaluateNamespaceName(resolutionCtx);
+
+        LOG.warn(
+            "The namespace '{}' of the workspace '{}' is not valid. Trying to recover"
+                + " from this situation using a default namespace which resolved to '{}'.",
+            namespace,
+            workspace.getId(),
+            defaultNamespace);
+
+        namespace = defaultNamespace;
+      } else {
+        // log a warning including a stacktrace to be able to figure out from where we got here...
+        LOG.warn(
```

Comment:
Will this case be hit during workspaces recovery?

Reply:
Nowadays, this should only be called during workspace delete, which I think is not called either from workspace recovery, nor from activity checker, which are the 2 jobs I know of that run without a valid subject.